libraries/chain/include/eosio/chain/trace.hpp

```diff
@@ -19,6 +19,63 @@ namespace eosio::chain {
    struct transaction_trace;
    using transaction_trace_ptr = std::shared_ptr<transaction_trace>;
 
+   struct sync_call_trace {
+      sync_call_trace(uint32_t sender_ordinal, account_name receiver, bool read_only, std::span<const char> data)
+         : sender_ordinal(sender_ordinal)
+         , receiver(receiver)
+         , read_only(read_only)
+         , data(data.begin(), data.end())
+      {
+      }
+
+      // receiver's ordinal. A sequence number starting with 1, unique within
+      // an action. It can be used to reference `call_trace` struct in `call_traces`
+      // vector in `action_trace` struct.
+      fc::unsigned_int              ordinal = 1;
+
+      // sender's ordinal. If the caller is the action, sender_ordinal is 0.
+      fc::unsigned_int              sender_ordinal = 0;
+
+      // receiver's account name
+      const account_name            receiver;
+
+      // indicate whether the call is read only or not.
+      const bool                    read_only = false;
+
+      // information about the call. Include function name, arguments, and other
+      // necessary information.
+      const std::vector<char>       data;
+
+      // the time the call takes
+      fc::microseconds              elapsed;
+
+      // the console log produced by the call
+      string                        console;
+
+      // for each call directly made by the current sync call, identify the starting
+      // position in `console`.
+      // For example, suppose current sync call:
+      //    prints 10 chars; calls sync1; prints 50 chars; calls sycn2,
+      // console_markers would look like { 10, 60 }
+      // This is used for pretty printing console logs to show hierachy of all logs.
+      std::vector<fc::unsigned_int>  console_markers;
```

Comment:
Should this be optional and not used if `contracts-console=false`? Or if not optional maybe it should be an empty vector if `contracts-console=false`.

Reply:
I think we should always provide it, as it can be used by `cleos` and other downstream tools to recreate call hierarchy for console logs.

Reply:
If `contracts-console=false` there is no console data. `console == ""`.

Reply:
This https://github.com/AntelopeIO/spring/blob/067730e131378cdb6f7c77a38cb179353a8b4f4e/libraries/chain/apply_context.cpp#L206 always executes, while `contracts-console` is only for immediate console printing:
https://github.com/AntelopeIO/spring/blob/067730e131378cdb6f7c77a38cb179353a8b4f4e/libraries/chain/apply_context.cpp#L191-L193

Reply:
That is going to be empty if `contracts-console=false`
https://github.com/AntelopeIO/spring/blob/7c352cedc2b63e9bb7f3926e442cf718aab428d5/libraries/chain/webassembly/console.cpp#L9-L9

Reply:
I think we keep it as an empty vector to match an empty console string if `contracts-console=false`?